// @storybook/addon-ondevice-knobs/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.register = register;
Object.defineProperty(exports, "withKnobs", {
  enumerable: true,
  get: function get() {
    return _addonKnobs.withKnobs;
  }
});

var _react = _interopRequireDefault(require("react"));

var _addons = _interopRequireDefault(require("@storybook/addons"));

var _panel = _interopRequireDefault(require("./panel"));

var _addonKnobs = require("@storybook/addon-knobs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function register() {
  _addons["default"].register('RNKNOBS', function () {
    var channel = _addons["default"].getChannel();

    _addons["default"].addPanel('RNKNOBS', {
      title: 'Knobs',
      // eslint-disable-next-line react/prop-types
      render: function render(_ref) {
        var active = _ref.active,
            key = _ref.key;
        return _react["default"].createElement(_panel["default"], {
          key: key,
          channel: channel,
          active: active
        });
      },
      paramKey: 'knobs'
    });
  });
}
